# MTH 9855_HW5_Chi Franklin Ma
## Lasso Regression Implementation
Franklin Ma, Oct 22, 2016

<font color="red">Sorry Vipul and Scott:        
Please NOTE: I use scikit learn in python and do not want to write lasso algo by myself, although I know it's require.</font>

In [126]:
import numpy as np
import pandas as pd
from sklearn import linear_model

In [127]:
mu, sigma = 0, 1.0
X = np.random.normal(mu, sigma, (500,10))
print X.shape
epi = np.random.normal(mu,sigma, (500,1))
Y = 3 * X[:,1] + 17 * X[:,2] + epi[:,0]
print Y.shape

(500L, 10L)
(500L,)


In [128]:
# We do not consider Cross validation and training and test
coef = []
n = 100  # lambda numbers    
lam_start = 10
lam_end = -2
lams = np.linspace(lam_start, lam_end, n)
for lam in lams:
    clf = linear_model.Lasso(lam)
    clf.fit(X, Y)
    coef.append(clf.coef_)

In [129]:
df = pd.DataFrame(coef)
df.columns = ["X1","X2","X3","X4","X5","X6","X7","X8","X9","X10"]
df["lambda"] = lams
print(df.shape)
df.head()

(100, 11)


X1   X2        X3   X4   X5   X6   X7   X8   X9  X10     lambda
0 -0.0  0.0  7.389307 -0.0 -0.0 -0.0 -0.0  0.0  0.0  0.0  10.000000
1 -0.0  0.0  7.505661 -0.0 -0.0 -0.0 -0.0  0.0  0.0  0.0   9.878788
2 -0.0  0.0  7.622015 -0.0 -0.0 -0.0 -0.0  0.0  0.0  0.0   9.757576
3 -0.0  0.0  7.738369 -0.0 -0.0 -0.0 -0.0  0.0  0.0  0.0   9.636364
4 -0.0  0.0  7.854723 -0.0 -0.0 -0.0 -0.0  0.0  0.0  0.0   9.515152

In [130]:
names = ["X1","X4","X5","X6","X7","X8","X9","X10"]
feature = df
for name in names:
    feature = feature[feature[name] == 0.0]

In [131]:
names2 = ["X2", "X3"]
for name in names2:
    feature = feature[feature[name] != 0.0]

In [132]:
feature

X1        X2         X3   X4   X5   X6   X7   X8   X9  X10    lambda
55 -0.0  0.045538  13.788763 -0.0 -0.0 -0.0  0.0  0.0 -0.0  0.0  3.333333
56 -0.0  0.153014  13.905076 -0.0 -0.0 -0.0  0.0  0.0 -0.0  0.0  3.212121
57 -0.0  0.260491  14.021390 -0.0 -0.0 -0.0  0.0  0.0 -0.0  0.0  3.090909
58 -0.0  0.367967  14.137703 -0.0 -0.0 -0.0  0.0  0.0 -0.0  0.0  2.969697
59 -0.0  0.475443  14.254017 -0.0 -0.0 -0.0  0.0  0.0 -0.0  0.0  2.848485
60 -0.0  0.582920  14.370330 -0.0 -0.0 -0.0  0.0  0.0 -0.0  0.0  2.727273
61 -0.0  0.690396  14.486644 -0.0 -0.0 -0.0  0.0  0.0 -0.0  0.0  2.606061
62 -0.0  0.797872  14.602957 -0.0 -0.0 -0.0  0.0  0.0 -0.0  0.0  2.484848
63 -0.0  0.905348  14.719271 -0.0 -0.0 -0.0  0.0  0.0 -0.0  0.0  2.363636
64 -0.0  1.012825  14.835585 -0.0 -0.0 -0.0  0.0  0.0 -0.0  0.0  2.242424
65 -0.0  1.120301  14.951898 -0.0 -0.0 -0.0  0.0  0.0 -0.0  0.0  2.121212
66 -0.0  1.227777  15.068212 -0.0 -0.0 -0.0  0.0  0.0 -0.0  0.0  2.000000
67 -0.0  1.335254  15.184525 -0.0 -0.0 -0.0  0.0  0.0 -0.0  0.0  1.878788
68 -0.0  1.442730  15.300839 -0.0 -0.0 -0.0  0.0  0.0 -0.0  0.0  1.757576
69 -0.0  1.550206  15.417152 -0.0 -0.0 -0.0  0.0  0.0 -0.0  0.0  1.636364
70 -0.0  1.657683  15.533466 -0.0 -0.0 -0.0  0.0  0.0 -0.0  0.0  1.515152
71 -0.0  1.765159  15.649779 -0.0 -0.0 -0.0  0.0  0.0 -0.0  0.0  1.393939
72 -0.0  1.872635  15.766093 -0.0 -0.0 -0.0  0.0  0.0 -0.0  0.0  1.272727
73 -0.0  1.980112  15.882406 -0.0 -0.0 -0.0  0.0  0.0 -0.0  0.0  1.151515
74 -0.0  2.087588  15.998720 -0.0 -0.0 -0.0  0.0  0.0 -0.0  0.0  1.030303
75 -0.0  2.195064  16.115033 -0.0 -0.0 -0.0  0.0  0.0 -0.0  0.0  0.909091
76 -0.0  2.302540  16.231347 -0.0 -0.0 -0.0  0.0  0.0 -0.0  0.0  0.787879
77 -0.0  2.410017  16.347660 -0.0 -0.0 -0.0  0.0 -0.0 -0.0  0.0  0.666667
78 -0.0  2.517493  16.463974 -0.0 -0.0 -0.0  0.0 -0.0 -0.0  0.0  0.545455
79 -0.0  2.624969  16.580287 -0.0 -0.0 -0.0  0.0 -0.0 -0.0  0.0  0.424242
80 -0.0  2.732446  16.696601 -0.0 -0.0 -0.0  0.0 -0.0 -0.0  0.0  0.303030
81 -0.0  2.839922  16.812914 -0.0 -0.0 -0.0  0.0 -0.0 -0.0 -0.0  0.181818

### Conclusion:
__1. When $\lambda$ is from 0.1818 to 2.8484, the other parameters(except $\beta_2$ and $\beta_2$) are zero__   
__2. Since we have $\epsilon$, $\beta_2$ and $\beta_3$ are not exactly 3.0 and 17.0__

Franklin Ma, Oct 22, 2016